### PR TITLE
AArch64: Implement read barrier evaluators

### PR DESCRIPTION
**Pull request by knn-k**
This commit implements read barrier evaluators for AArch64.

Signed-off-by: KONNO Kazuhiro <konno@jp.ibm.com>

### CODE REVIEW

**Comment by knn-k** [comment]
jenkins compile alinux64 jdk11

**Comment by knn-k** [comment]
Related PR: https://github.com/eclipse/omr/pull/5224

**Comment by knn-k** [comment]
Ready for review.  #9665 depends on this PR.
@0xdaryl FYI.

**Comment by knn-k** [comment]
The Travis CI build failure was not caused by this PR.
```
[ 83%] Building ASM_NASM object runtime/compiler/CMakeFiles/j9jit.dir/x/runtime/X86ArrayTranslate.nasm.o
nasm: error: invalid warning `return-type'
type `nasm -h' for help
runtime/compiler/CMakeFiles/j9jit.dir/build.make:2287: recipe for target 'runtime/compiler/CMakeFiles/j9jit.dir/x/runtime/X86ArrayTranslate.nasm.o' failed
```

**Comment by knn-k** [comment]
jenkins compile alinux64 jdk11

**Comment by knn-k** [comment]
Try running system tests before enabling concurrent scavenger.

**Comment by knn-k** [comment]
jenkins test sanity.system alinux64 jdk11

**Comment by knn-k** [comment]
The failure with TestIBMJlmRemoteMemoryAuth_0 is a known issue #9047.

**Comment by knn-k** [comment]
I manually ran the following tests for concurrent scavenger from openjdk-tests/system on cent7-aarch64-1 with my local build, and they all passed.

- DaaLoadTest_daa1_ConcurrentScavenge
- LambdaLoadTest_ConcurrentScavenge
- MathLoadTest_bigdecimal_ConcurrentScavenge
- MauveSingleThreadLoadTest_ConcurrentScavenge
- ClassLoadingTest_ConcurrentScavenge

**Comment by knn-k** [comment]
Minor update for avoiding compiler warnings.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -110,6 +111,286 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
...
+   cg->stopUsingRegister(x0Reg);
+
+   cg->machine()->setLinkRegisterKilled(true);
+
+   return objReg;
+#endif // OMR_GC_CONCURRENT_SCAVENGER
+   }
+
+static TR::Register *
+aGenerateSoftwareReadBarrier(TR::Node *node, TR::CodeGenerator *cg)
```
Comment (0xdaryl) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:209]: The DNA of this method is about 80-90% identical to the iGenerateSoftwareReadBarrier method.  Is there any way you could combine them to save some code space?

Comment (0xdaryl) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:209]: Most of my comments in iGenerateSoftwareReadBarrier apply to this function as well so I'm not going to repeat them.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -110,6 +111,286 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
...
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, endLabel, deps);
+
+   bool needSync = (node->getSymbolReference()->getSymbol()->isSyncVolatile() && cg->comp()->target().isSMP());
+
+   if (needSync)
+      {
+      generateSynchronizationInstruction(cg, TR::InstOpCode::dsb, node, 0xF); // dsb SY
+      }
+
+   cg->insertPrefetchIfNecessary(node, objReg);
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:194]: It came from the p version.  Removing.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -110,6 +111,286 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
...
+   TR::LabelSymbol *endLabel = generateLabelSymbol(cg);
+   startLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+   TR::RegisterDependencyConditions *deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 5, cg->trMemory());
+   deps->addPostCondition(objReg, TR::RealRegister::NoReg);
+   deps->addPostCondition(locationReg, TR::RealRegister::x1); // TR_softwareReadBarrier helper needs this in x1.
+   deps->addPostCondition(evacuateReg, TR::RealRegister::NoReg);
+   deps->addPostCondition(x0Reg, TR::RealRegister::x0);
+   deps->addPostCondition(metaReg, TR::RealRegister::NoReg);
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:145]: It came from the p version.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 4]
```diff
@@ -110,6 +111,286 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
...
+
+   tempMR = new (cg->trHeapMemory()) TR::MemoryReference(node, 4, cg);
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:147]: It is not an offset but a length parameter.
It is meaningless, and is addressed by https://github.com/eclipse/omr/issues/4137 .

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 5]
```diff
@@ -110,6 +111,286 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
...
+      else
+         generateTrg1MemInstruction(cg, TR::InstOpCode::addimmx, node, locationReg, tempMR);
+      }
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, objReg, new (cg->trHeapMemory()) TR::MemoryReference(locationReg, 0, cg));
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel);
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, evacuateReg,
+         new (cg->trHeapMemory()) TR::MemoryReference(metaReg, cg->comp()->fej9()->thisThreadGetEvacuateBaseAddressOffset(), cg));
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Changed to `comp`, here and other locations.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 1 above)
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Merged the two functions into `generateSoftwareReadBarrier()`, which takes `bool isArdbari` as its third argument.

**Comment by knn-k** [comment]
Reflected the review comments.

**Comment by knn-k** [comment]
jenkins compile alinux64 jdk11

**Comment by knn-k** [comment]
I ran some of the load tests (mentioned above) with both compressed refs and large heap builds, and saw no failures.

**Pull request merged**